src/entropic/sinkhorn.jl

```diff
@@ -209,3 +222,113 @@ function sinkhorn2(μ, ν, C, ε, alg::Sinkhorn; regularization=false, plan=noth
+Compute the Sinkhorn Divergence between finite discrete
+measures `μ` and `ν` with respect to a common cost matrix `C`,
+entropic regularization parameter `ε` and algorithm `alg`. 
+
+For `regularization = true`, the Sinkhorn Divergence is that of [^FeydyP19] and is computed as:
+```math
+\\operatorname{S}_{ε}(μ,ν) := \\operatorname{OT}_{ε}(μ,ν)
+- \\frac{1}{2}(\\operatorname{OT}_{ε}(μ,μ) + \\operatorname{OT}_{ε}(ν,ν)),
+```
+where ``\\operatorname{OT}_{ε}(μ,ν)``, ``\\operatorname{OT}_{ε}(μ,μ)`` and
+``\\operatorname{OT}_{ε}(ν,ν)`` are the entropically regularized optimal transport loss 
+between `(μ,ν)`, `(μ,μ)` and `(ν,ν)`, respectively.
+
+For `regularization = false`, the Sinkhorn Divergence is that of [^GPC18] and is computed as:
+```math
+\\operatorname{S}_{ε}(μ,ν) := \\operatorname{W}_{ε}(μ,ν)
+- \\frac{1}{2}(\\operatorname{W}_{ε}(μ,μ) + \\operatorname{W}_{ε}(ν,ν)),
+```
+where ``\\operatorname{W}_{ε}(μ,ν)``, ``\\operatorname{W}_{ε}(μ,μ)`` and
+``\\operatorname{W}_{ε}(ν,ν)`` are the entropically regularized optimal transport *cost*
+between `(μ,ν)`, `(μ,μ)` and `(ν,ν)`, respectively, defined as 
+```math
+\\operatorname{W}_{ε}(μ, ν) = \\langle C, γ^\\star \\rangle
+```
+where ``γ^\\star`` is the entropy-regularised transport plan between `μ` and `ν`. 
+The default algorithm for computing the term ``\\operatorname{OT}_{ε}(μ, ν)`` is the `SinkhornGibbs` algorithm.
+For the terms ``\\operatorname{OT}_{ε}(μ, μ)`` and ``\\operatorname{OT}_{ε}(ν, ν)``, the symmetric fixed point iteration of [^FeydyP19] is used. 
+Alternatively, a pre-computed optimal transport `plan` between `μ` and `ν` may be provided.  
+
+[^GPC18]: Aude Genevay, Gabriel Peyré, Marco Cuturi, Learning Generative Models with Sinkhorn Divergences, Proceedings of the Twenty-First International Conference on Artficial Intelligence and Statistics, (AISTATS) 21, 2018
+[^FeydyP19]: Jean Feydy, Thibault Séjourné, François-Xavier Vialard, Shun-ichi Amari, Alain Trouvé, and Gabriel Peyré. Interpolating between optimal transport and mmd using sinkhorn divergences. In The 22nd International Conference on Artificial Intelligence and Statistics, pages 2681–2690. PMLR, 2019.
+See also: [`sinkhorn2`](@ref)
+"""
+function sinkhorn_divergence(
+    μ::AbstractVecOrMat,
+    ν::AbstractVecOrMat,
+    C,
+    ε,
+    alg::Sinkhorn=SinkhornGibbs(),
+    algμ::Sinkhorn=SymmetricSinkhornGibbs(),
+    algν::Sinkhorn=SymmetricSinkhornGibbs();
+    regularization=true,
+    plan=nothing,
+    kwargs...,
+)
+    OTμν, OTμ, OTν = if (regularization == true) && (plan === nothing)
+        sinkhorn_loss(μ, ν, C, ε, alg; kwargs...),
+        sinkhorn_loss(μ, C, ε, algμ; kwargs...),
+        sinkhorn_loss(ν, C, ε, algν; kwargs...)
+    else
+        sinkhorn2(μ, ν, C, ε, alg; plan=plan, regularization=false, kwargs...),
+        sinkhorn2(μ, C, ε, algμ; plan=nothing, regularization=false, kwargs...),
+        sinkhorn2(ν, C, ε, algν; plan=nothing, regularization=false, kwargs...)
+    end
```

Comment:
Again, can we handle this in the other functions?
```suggestion
    OTμν = sinkhorn2(μ, ν, C, ε, alg; regularization=regularization, kwargs...)
    OTμ = sinkhorn2(μ, C, ε, algμ; regularization=regularization, kwargs...)
    OTν = sinkhorn2(ν, C, ε, algν; regularization=regularization, kwargs...)
```

Reply:
This was what I was hoping to accomplish initially. However, the considerations I had in mind were:
- `sinkhorn2` currently computes the transport plan matrix, whereas `sinkhorn_loss` only uses the dual variables `(u, v)`. So the internal implementations would need to be diffetagrent.
- We can only use the `sinkhorn_dual_obj` method when `regularization = true`. It would be nice to pass tags such as `DualObjective(), Cost()` etc and dispatch to different methods. 

If we want to handle it in `sinkhorn2`, then I think the quick fix will be to handle the special case `regularization == true && plan === nothing` inside. I guess both solutions for now are quite ugly.

Reply:
> If we want to handle it in `sinkhorn2`, then I think the quick fix will be to handle the special case `regularization == true && plan === nothing` inside as an if-else block. I guess both solutions for now are quite ugly.

I think we should do this nevertheless. IMO it will be more readable and a cleaner API if there is only one function. I assume we could also use a similar approach for the case `regularization = false` and `plan = nothing` and avoid having to instantiate the plan?

Reply:
Yeah sounds good. One additional consideration is that it should only be used for `SinkhornGibbs` algorithm. I've implemented this now.